File: app/src/main/java/com/example/ttools/MainActivity.java
package com.example.ttools;

import android.content.Intent;
import android.os.Bundle;

import com.bumptech.glide.Glide;
import com.example.ttools.APISERVER.models.ApiNews;
import com.example.ttools.APISERVER.TibiaAPIServer;
import com.example.ttools.APISERVER.models.APICriatures;
import com.example.ttools.APISERVER.models.ApiNewsTicker;
import com.example.ttools.APISERVER.models.criatures.BoostableBosses;
import com.example.ttools.APISERVER.models.criatures.Criatures;
import com.example.ttools.Operaciones.InstanciaRetrofit;
import com.example.ttools.recyclerview.Adapters.AdapterRecyclerViewNews;
import com.example.ttools.recyclerview.ItemsRecyclerViewNews;
import com.example.ttools.utilidades.RedValidator;
import com.google.android.material.navigation.NavigationView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity implements View.OnClickListener, MenuItem.OnMenuItemClickListener {
    /**
     * Navigation Drawer
     * */
    DrawerLayout drawerLayout;
    ActionBarDrawerToggle actionBarDrawerToggle;
    NavigationView navigationView;

    ImageView imgRashid, imgBossCreature, imgBossBosstable;
    TextView lbRashid, textBossCreature,lbBossBosstable;
    //para las new latest
    TextView textViewFecha,textViewCategoria, textViewNoticia,textViewTipo;
    TextView textViewDate, textViewCategory, textViewNew, textViewtype;
    String url = "https://api.tibialabs.com/v2/";
    String url_rashid_image = "https://raw.githubusercontent.com/MiguelJeronimo/TtoolsDesktop/main/src/img/rashid.gif";
    String url_creature_boss = "https://api.tibiadata.com/v3/";
    String url_boosted_boss = "https://api.tibiadata.com/v3/";
    String url_new = "https://api.tibiadata.com/v3/";
    //recyclerview
    RecyclerView recyclerViewNoticas;
    AdapterRecyclerViewNews adapterRecyclerViewNews;
    List<ItemsRecyclerViewNews> itemsRecyclerViewNewsList;
    //RedValidator redValidator = new RedValidator();
    //retrofit
    InstanciaRetrofit servicio = new InstanciaRetrofit();
    //Asincronia asincronia = new Asincronia();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        drawerLayout = findViewById(R.id.navegacion);
        //Agrega utilidades a la toolbar, asi como el incono de hamburguesa, y los iconos de cuando se despliegue el menu y cuando este cerrado
        actionBarDrawerToggle = new ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.nav_open,R.string.nav_close);
        //agrega el evento al drawelayout, recibe por parametro el actionBarDrawerToggle en la cual se encuentra las animaciones
        drawerLayout.addDrawerListener(actionBarDrawerToggle);
        //sincroniza los estados del navigationDrawer
        actionBarDrawerToggle.syncState();
        navigationView = findViewById(R.id.barraNavegacion);
        //a cada item del menuo agregamos su evento MenuItemClickListener
        navigationView.getMenu().findItem(R.id.nd_character).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.nd_criaturas).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.nd_stamina).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.nd_experiencia_compartida).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.nd_mundos).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.nd_guilds).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.nd_spells).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.nd_bless).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.nd_highscores).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.nd_houses).setOnMenuItemClickListener(this);
        navigationView.getMenu().findItem(R.id.id_about).setOnMenuItemClickListener(this);

        imgRashid = findViewById(R.id.imageViewRashid);
        lbRashid = findViewById(R.id.textViewRashid);
        imgBossCreature = findViewById(R.id.BossCreature);
        textBossCreature = findViewById(R.id.textBossCreature);
        imgBossBosstable = findViewById(R.id.BossBosstable);
        lbBossBosstable = findViewById(R.id.lbBossBosstable);
        //noticias
        textViewFecha = findViewById(R.id.textViewFecha);
        textViewCategoria = findViewById(R.id.textViewCategoria);
        textViewNoticia = findViewById(R.id.textViewNoticia);
        textViewTipo = findViewById(R.id.textViewTipo);
        textViewDate = findViewById(R.id.textViewDate);
        textViewCategory = findViewById(R.id.textViewCategory);
        textViewNew = findViewById(R.id.textViewNew);
        textViewtype = findViewById(R.id.textViewtype);
        //Validamos si tenenemos conexion a internet
        if (RedValidator.ValidarInternet(this)){
            //ejecutando los multiple hilos para el consumo de api
            //asincronia.execute();
            Hilos();
        } else{
            Toast.makeText(this,"Revisa tu conexion a internet :)",Toast.LENGTH_SHORT).show();
        }

    }

    public void llenarRashid(String url){
        TibiaAPIServer tibiaAPIServer = servicio.getRetrofit(url).create(TibiaAPIServer.class);
        Call<String> call = tibiaAPIServer.getRashidLocalitation();
        call.enqueue(new Callback<String>() {
            @Override
            public void onResponse(Call<String> call, Response<String> response) {
                if (response.isSuccessful()){
                    lbRashid.setText(response.body());
                    Glide.with(getApplicationContext()).load(url_rashid_image).asGif().into(imgRashid);
                }
            }

            @Override
            public void onFailure(Call<String> call, Throwable t) {
                System.out.println("MENSAJE"+t.getMessage());
            }
        });
    }

    public void creatureBoss(String url_creature_boss){
        TibiaAPIServer tibiaAPIServer =  servicio.getRetrofit(url_creature_boss).create(TibiaAPIServer.class);
        Call<APICriatures> call = tibiaAPIServer.getCreatures();
        call.enqueue(new Callback<APICriatures>() {
            @Override
            public void onResponse(Call<APICriatures> call, Response<APICriatures> response) {
                if (!response.isSuccessful()) {
                    return;
                }
                APICriatures apiCriatures = response.body();
                assert apiCriatures != null;
                Criatures criatures = apiCriatures.getCreatures();
                Glide.with(getApplicationContext()).load(criatures.getBoosted().getImage_url()).asGif().into(imgBossCreature);
                textBossCreature.setText(criatures.getBoosted().getName());

            }

            @Override
            public void onFailure(Call<APICriatures> call, Throwable t) {
                System.out.println("ERROR: "+t.getMessage());
            }
        });
    }
    public void bostedBoss(String url_boosted_boss){
        TibiaAPIServer apiServer = servicio.getRetrofit(url_boosted_boss).create(TibiaAPIServer.class);
        Call<APICriatures> call = apiServer.getBooted();
        call.enqueue(new Callback<APICriatures>() {
            @Override
            public void onResponse(Call<APICriatures> call, Response<APICriatures> response) {
                if (response.isSuccessful()){
                    APICriatures apiCriatures = response.body();
                    assert apiCriatures != null;
                    BoostableBosses boostableBosses = apiCriatures.getBoostable_bosses();
                    Glide.with(getApplicationContext()).load(boostableBosses.getBoosted().getImage_url()).asGif().into(imgBossBosstable);
                    lbBossBosstable.setText(boostableBosses.getBoosted().getName());
                }
            }

            @Override
            public void onFailure(Call<APICriatures> call, Throwable t) {
                System.out.println(t.getMessage());
            }
        });
    }

    public void Noticas(String url_new){
        TibiaAPIServer tibiaAPIServer = servicio.getRetrofit(url_new).create(TibiaAPIServer.class);
        Call<ApiNews> call = tibiaAPIServer.getNewsLatest();
        call.enqueue(new Callback<ApiNews>() {
            @Override
            public void onResponse(Call<ApiNews> call, Response<ApiNews> response) {
                if (response.isSuccessful()){
                   ApiNews apiNews = response.body();
                   //Primer Card
                   textViewFecha.setText(apiNews.getNews().get(0).getDate());
                   textViewCategoria.setText(apiNews.getNews().get(0).getCategory());
                   textViewNoticia.setText(apiNews.getNews().get(0).getNews());
                   textViewTipo.setText(apiNews.getNews().get(0).getType());
                   //Segundo Card
                   textViewDate.setText(apiNews.getNews().get(1).getDate());
                   textViewCategory.setText(apiNews.getNews().get(1).getCategory());
                   textViewNew.setText(apiNews.getNews().get(1).getNews());
                   textViewtype.setText(apiNews.getNews().get(1).getType());
                }
            }

            @Override
            public void onFailure(Call<ApiNews> call, Throwable t) {
                System.out.println(t.getMessage());
            }
        });
    }

    public void NewsTickers(String url_new){
        recyclerViewNoticas = findViewById(R.id.recyclerNews);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        TibiaAPIServer tibiaAPIServer = servicio.getRetrofit(url_new).create(TibiaAPIServer.class);
        Call<ApiNewsTicker> call = tibiaAPIServer.getNewsTickers();
        call.enqueue(new Callback<ApiNewsTicker>() {
            @Override
            public void onResponse(Call<ApiNewsTicker> call, Response<ApiNewsTicker> response) {
                if (response.isSuccessful()){
                    ApiNewsTicker apiNewsTicker = response.body();
                    itemsRecyclerViewNewsList = new ArrayList<>();
                    for (int i=0; i<5; i++){
                        itemsRecyclerViewNewsList.add(new ItemsRecyclerViewNews(
                                apiNewsTicker.getNews().get(i).getId(),
                                apiNewsTicker.getNews().get(i).getDate(),
                                apiNewsTicker.getNews().get(i).getNews(),
                                apiNewsTicker.getNews().get(i).getCategory(),
                                apiNewsTicker.getNews().get(i).getType(),
                                apiNewsTicker.getNews().get(i).getUrl()
                        ));
                    }
                    recyclerViewNoticas.setLayoutManager(layoutManager);
                    adapterRecyclerViewNews = new AdapterRecyclerViewNews(itemsRecyclerViewNewsList);
                    layoutManager.setOrientation(LinearLayoutManager.HORIZONTAL);
                    //recyclerViewNoticas.setHasFixedSize(true);
                    recyclerViewNoticas.setAdapter(adapterRecyclerViewNews);
                }
            }

            @Override
            public void onFailure(@NonNull Call<ApiNewsTicker> call, Throwable t) {
                System.out.println(t.getMessage());
            }
        });
    }
    public void Hilos(){
        Thread hilo = new Thread(new Runnable() {
            @Override
            public void run() {
                llenarRashid(url);
            }
        });
        Thread hilo2 = new Thread(new Runnable() {
            @Override
            public void run() {
                creatureBoss(url_creature_boss);
            }
        });
        Thread hilo3 = new Thread(new Runnable() {
            @Override
            public void run() {
                bostedBoss(url_boosted_boss);
            }
        });

        Thread hilo4 = new Thread(new Runnable() {
            @Override
            public void run() {
                Noticas(url_new);
            }
        });
        Thread hilo5 = new Thread(new Runnable() {
            @Override
            public void run() {
                NewsTickers(url_new);
            }
        });
        hilo.start();
        hilo2.start();
        hilo3.start();
        hilo4.start();
        hilo5.start();
    }

//implementacion de tareas asincronas
/*private class Asincronia extends AsyncTask <String, Void, String> {
    Thread hilo = new Thread(new Runnable() {
        @Override
        public void run() {
            llenarRashid(url);
        }
    });
    Thread hilo2 = new Thread(new Runnable() {
        @Override
        public void run() {
            creatureBoss(url_creature_boss);
        }
    });
    Thread hilo3 = new Thread(new Runnable() {
        @Override
        public void run() {
            bostedBoss(url_boosted_boss);
        }
    });

    Thread hilo4 = new Thread(new Runnable() {
        @Override
        public void run() {
            Noticas(url_new);
        }
    });
    Thread hilo5 = new Thread(new Runnable() {
        @Override
        public void run() {
            NewsTickers(url_new);
        }
    });

    @Override
    protected String doInBackground(String... strings) {
        hilo.start();
        hilo2.start();
        hilo3.start();
        hilo4.start();
        hilo5.start();
        return null;
    }
}*/
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();
        //noinspection SimplifiableIfStatement
        if (id == R.id.nd_about){
            Intent about = new Intent(this, About.class);
            startActivity(about);
        } else if (id == R.id.nd_salir){
            finish();
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onClick(View v) {

    }

    @Override
    public boolean onMenuItemClick(MenuItem menuItem) {

        if (menuItem.getItemId() == R.id.nd_experiencia_compartida){
            Intent expcompartida = new Intent(this, experiencia_compartida.class);
            startActivity(expcompartida);
            drawerLayout.close();
        } else if (menuItem.getItemId() == R.id.nd_character){
            Intent characters = new Intent(this, characters.class);
            startActivity(characters);
            drawerLayout.close();
        } else if (menuItem.getItemId() == R.id.nd_stamina){
            Intent stamina = new Intent(this, Stamina.class);
            startActivity(stamina);
            drawerLayout.close();
        }else if (menuItem.getItemId() == R.id.nd_mundos){
            Intent mundos = new Intent(this, Mundos.class);
            startActivity(mundos);
            drawerLayout.close();
        }else if(menuItem.getItemId() == R.id.nd_criaturas){
            Intent criaturas = new Intent(this, Criaturas.class);
            startActivity(criaturas);
            drawerLayout.close();
        } else if (menuItem.getItemId() == R.id.nd_guilds){
            Intent guilds = new Intent(this, GuildInformation.class);
            startActivity(guilds);
            drawerLayout.close();
        } else if (menuItem.getItemId() == R.id.nd_spells){
            Intent spellsTibia = new Intent(this, Spells_Tibia.class);
            startActivity(spellsTibia);
            drawerLayout.close();
        }else if (menuItem.getItemId() == R.id.nd_bless){
            Intent blessTibia= new Intent(this, Blessings.class);
            startActivity(blessTibia);
            drawerLayout.close();
        } else if(menuItem.getItemId() == R.id.nd_highscores){
            Intent HighscoresTibia= new Intent(this, Highscores.class);
            startActivity(HighscoresTibia);
            drawerLayout.close();
        } else if (menuItem.getItemId() == R.id.nd_houses){
            Intent tibiaHouses = new Intent(this, HouseActivity.class);
            startActivity(tibiaHouses);
            drawerLayout.close();
        } else if(menuItem.getItemId() == R.id.id_about){
            Intent about = new Intent(this, About.class);
            startActivity(about);
            drawerLayout.close();
        }
        return false;
    }
}

File: app/src/main/java/com/example/TibiaTools/MainActivity.kt
package com.example.TibiaTools

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.appcompat.widget.Toolbar
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.TibiaTools.data.model.*
import com.example.TibiaTools.View.About
import com.example.TibiaTools.View.Blessings
import com.example.TibiaTools.View.Criaturas
import com.example.TibiaTools.View.GuildInformation
import com.example.TibiaTools.View.Highscores
import com.example.TibiaTools.View.HouseActivity
import com.example.TibiaTools.View.Mundos
import com.example.TibiaTools.View.Spells_Tibia
import com.example.TibiaTools.View.Stamina
import com.example.TibiaTools.View.TibiaMaps
import com.example.TibiaTools.View.ViewModel.ViewModelHome
import com.example.TibiaTools.View.characters
import com.example.TibiaTools.View.experiencia_compartida
import com.example.TibiaTools.recyclerview.Adapters.AdapterRecyclerViewNews
import com.example.TibiaTools.recyclerview.ItemsRecyclerViewNews
import com.example.TibiaTools.utilidades.RedValidator
import com.example.ttools.R
import com.github.AAChartModel.AAChartCore.AAChartCreator.AAChartModel
import com.github.AAChartModel.AAChartCore.AAChartCreator.AAChartView
import com.github.AAChartModel.AAChartCore.AAChartCreator.AASeriesElement
import com.github.AAChartModel.AAChartCore.AAChartEnum.AAChartType
import com.github.AAChartModel.AAChartCore.AAOptionsModel.AAStyle
import com.google.android.material.card.MaterialCardView
import com.google.android.material.navigation.NavigationView
import com.google.android.material.progressindicator.LinearProgressIndicator
import org.koin.android.ext.android.inject
import java.util.ArrayList
import java.util.Collections
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

class MainActivity : AppCompatActivity(), View.OnClickListener, MenuItem.OnMenuItemClickListener {
    private lateinit var drawerLayout: DrawerLayout
    private lateinit var actionBarDrawerToggle: ActionBarDrawerToggle
    private lateinit var navigationView: NavigationView

    private lateinit var imgRashid: ImageView
    private lateinit var imgBossCreature: ImageView
    private lateinit var imgBossBosstable: ImageView
    private lateinit var lbRashid: TextView
    private lateinit var textBossCreature: TextView
    private lateinit var lbBossBosstable: TextView

    private lateinit var textViewFecha: TextView
    private lateinit var textViewCategoria: TextView
    private lateinit var textViewNoticia: TextView
    private lateinit var textViewTipo: TextView
    private lateinit var textViewDate: TextView
    private lateinit var textViewCategory: TextView
    private lateinit var textViewNew: TextView
    private lateinit var textViewtype: TextView
    private lateinit var textViewPlayesOnline: TextView

    private lateinit var recyclerViewNoticas: RecyclerView
    private lateinit var adapterRecyclerViewNews: AdapterRecyclerViewNews
    private lateinit var linearProgressIndicator: LinearProgressIndicator
    private lateinit var itemsRecyclerViewNewsList: MutableList<ItemsRecyclerViewNews>

    private lateinit var cardCreatureBooss: MaterialCardView
    private lateinit var cardBoostedBoos: MaterialCardView
    private lateinit var cardNews: MaterialCardView
    private lateinit var cardNews2: MaterialCardView

    private val viewModelHome: ViewModelHome by inject()
    private lateinit var aaChartView: AAChartView
    private lateinit var aaChartView2: AAChartView
    private val data = ArrayList<Double>()

    @RequiresApi(Build.VERSION_CODES.N)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        cardNews = findViewById(R.id.cardView4)
        cardNews2 = findViewById(R.id.cardView5)
        drawerLayout = findViewById(R.id.navegacion)
        cardCreatureBooss = findViewById(R.id.cardCreatureBoos)
        cardBoostedBoos = findViewById(R.id.cardBoostedBoos)

        actionBarDrawerToggle = ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.nav_open, R.string.nav_close)
        drawerLayout.addDrawerListener(actionBarDrawerToggle)
        actionBarDrawerToggle.syncState()

        linearProgressIndicator = findViewById(R.id.carga_main)
        navigationView = findViewById(R.id.barraNavegacion)

        val menu = navigationView.menu
        menu.findItem(R.id.nd_character)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_criaturas)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_stamina)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_experiencia_compartida)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_mundos)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_guilds)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_spells)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_bless)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_highscores)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_houses)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.id_about)?.setOnMenuItemClickListener(this)
        menu.findItem(R.id.nd_maps)?.setOnMenuItemClickListener(this)

        imgRashid = findViewById(R.id.imageViewRashid)
        lbRashid = findViewById(R.id.textViewRashid)
        imgBossCreature = findViewById(R.id.BossCreature)
        textBossCreature = findViewById(R.id.textBossCreature)
        imgBossBosstable = findViewById(R.id.BossBosstable)
        lbBossBosstable = findViewById(R.id.lbBossBosstable)

        textViewFecha = findViewById(R.id.textViewFecha)
        textViewCategoria = findViewById(R.id.textViewCategoria)
        textViewNoticia = findViewById(R.id.textViewNoticia)
        textViewTipo = findViewById(R.id.textViewTipo)
        textViewDate = findViewById(R.id.textViewDate)
        textViewCategory = findViewById(R.id.textViewCategory)
        textViewNew = findViewById(R.id.textViewNew)
        textViewtype = findViewById(R.id.textViewtype)
        textViewPlayesOnline = findViewById(R.id.players_online_main)

        val timer = Timer()
        recyclerViewNoticas = findViewById(R.id.recyclerNews)
        itemsRecyclerViewNewsList = ArrayList()
        val layoutManager = LinearLayoutManager(this)
        recyclerViewNoticas.layoutManager = layoutManager
        adapterRecyclerViewNews = AdapterRecyclerViewNews(itemsRecyclerViewNewsList)
        layoutManager.orientation = LinearLayoutManager.HORIZONTAL
        recyclerViewNoticas.setHasFixedSize(true)
        recyclerViewNoticas.adapter = adapterRecyclerViewNews

        aaChartView = findViewById(R.id.AAChartView)
        aaChartView2 = findViewById(R.id.AAChartView2)
        val aaChartModel = AAChartModel()
        val aaChartModel2 = AAChartModel()

        viewModelHome.rashidLocation.observe(this) { location ->
            if (location != null) {
                val urlRashidImage = "https://raw.githubusercontent.com/MiguelJeronimo/TtoolsDesktop/main/src/img/rashid.gif"
                lbRashid.text = location
                Glide.with(applicationContext).load(urlRashidImage).into(imgRashid)
                imgRashid.visibility = View.VISIBLE
                lbRashid.visibility = View.VISIBLE
            } else {
                imgRashid.visibility = View.GONE
                lbRashid.visibility = View.GONE
            }
        }

        val dataElement = ArrayList<AASeriesElement>()
        val categories = ArrayList<String>()
        val data2 = ArrayList<Double>()
        val isDrawChart2 = AtomicBoolean(false)

        viewModelHome.worlds().observe(this) { worlds ->
            if (worlds != null) {
                categories.clear()
                dataElement.clear()
                data2.clear()
                val worldsList = ArrayList<RegularWorlds>(worlds.regular_worlds)
                worldsList.sortWith(Collections.reverseOrder())
                for (world in worldsList) {
                    categories.add(world.name ?: "")
                    data2.add(world.players_online.toDouble())
                }
                dataElement.add(
                    AASeriesElement()
                        .name("Onlines")
                        .data(data2.toTypedArray())
                )
                val serieelement = dataElement.toTypedArray()
                val categoriesArray = categories.toTypedArray()
                if (isDrawChart2.get()) {
                    aaChartView2.aa_onlyRefreshTheChartDataWithChartOptionsSeriesArray(serieelement)
                } else {
                    aaChartModel2.chartType(AAChartType.Bar)
                        .animationType("Bounce")
                        .title("Worlds").titleStyle(AAStyle().color("#DCE4E9"))
                        .subtitle("Players online to worlds in Tibia").subtitleStyle(AAStyle().color("#70787D"))
                        .backgroundColor("#191C1E")
                        .dataLabelsEnabled(false)
                        .categories(categoriesArray)
                        .yAxisGridLineWidth(0f)
                        .colorsTheme(arrayOf("#67D3FF"))
                        .series(dataElement.toTypedArray())
                    aaChartView2.aa_drawChartWithChartModel(aaChartModel2)
                    aaChartView2.visibility = View.VISIBLE
                    isDrawChart2.set(true)
                }
            }
        }

        val isDrawChart = AtomicBoolean(false)
        viewModelHome.playersOnline.observe(this) { playersOnline ->
            if (playersOnline != null) {
                val onlines = "Players Online: ${playersOnline.players_online}"
                textViewPlayesOnline.visibility = View.VISIBLE
                textViewPlayesOnline.text = onlines
                val players = playersOnline.players_online.toDouble()
                if (data.isNotEmpty()) {
                    val position = data.size - 1
                    if (data[position] != players) {
                        data.add(players)
                    }
                } else {
                    data.add(players)
                }
                aaChartModel.chartType(AAChartType.Spline)
                    .animationType("Bounce")
                    .title("PLAYERS ONLINE").titleStyle(AAStyle().color("#DCE4E9"))
                    .subtitle("Players online in Tibia").subtitleStyle(AAStyle().color("#70787D"))
                    .backgroundColor("#191C1E")
                    .dataLabelsEnabled(false)
                    .yAxisGridLineWidth(0f)

                if (isDrawChart.get()) {
                    val dataElementChart = arrayOf(
                        AASeriesElement()
                            .name("Onlines")
                            .data(data.toTypedArray())
                    )
                    aaChartView.aa_onlyRefreshTheChartDataWithChartOptionsSeriesArray(dataElementChart)
                } else {
                    aaChartModel.series(
                        arrayOf(
                            AASeriesElement()
                                .name("Onlines")
                                .data(data.toTypedArray())
                        )
                    )
                    aaChartView.aa_drawChartWithChartModel(aaChartModel)
                    aaChartView.visibility = View.VISIBLE
                    isDrawChart.set(true)
                }
            } else {
                textViewPlayesOnline.visibility = View.GONE
                textViewPlayesOnline.text = ""
            }
        }

        viewModelHome.creatureBoss.observe(this) { creatureBoss ->
            if (creatureBoss != null) {
                Glide.with(applicationContext).load(creatureBoss.boosted?.image_url).into(imgBossCreature)
                textBossCreature.text = creatureBoss.boosted?.name
                cardCreatureBooss.visibility = View.VISIBLE
            } else {
                linearProgressIndicator.visibility = View.GONE
                cardCreatureBooss.visibility = View.GONE
            }
        }

        viewModelHome.boostedBoss.observe(this) { bostedBoss ->
            if (bostedBoss != null) {
                val boosted = bostedBoss.boosted
                Glide.with(applicationContext).load(boosted?.image_url).into(imgBossBosstable)
                lbBossBosstable.text = boosted?.name
                cardBoostedBoos.visibility = View.VISIBLE
            } else {
                cardBoostedBoos.visibility = View.GONE
            }
        }

        viewModelHome.news.observe(this) { news ->
            if (news != null) {
                val sizeNews = news.news.size
                val rangeNews = Math.min(sizeNews, 2)
                if (sizeNews > 0) {
                    for (i in 0 until rangeNews) {
                        textViewFecha.text = news.news[i].date
                        textViewCategoria.text = news.news[i].category
                        textViewNoticia.text = news.news[i].news
                        textViewTipo.text = news.news[i].type
                        // Segundo Card
                        textViewDate.text = news.news[i + 1].date
                        textViewCategory.text = news.news[i + 1].category
                        textViewNew.text = news.news[i + 1].news
                        textViewtype.text = news.news[i + 1].type
                        cardNews.visibility = View.VISIBLE
                        cardNews2.visibility = View.VISIBLE
                    }
                }
            } else {
                cardNews.visibility = View.GONE
                cardNews2.visibility = View.GONE
            }
        }

        viewModelHome.newTicker.observe(this) { newTicker ->
            if (newTicker != null) {
                newTicker.news.forEach { news ->
                    itemsRecyclerViewNewsList.add(
                        ItemsRecyclerViewNews(
                            news.id,
                            news.date,
                            news.news,
                            news.category,
                            news.type,
                            news.url
                        )
                    )
                }
                adapterRecyclerViewNews.notifyDataSetChanged()
                linearProgressIndicator.visibility = View.GONE
            } else {
                linearProgressIndicator.visibility = View.GONE
            }
        }

        timer.schedule(object : TimerTask() {
            override fun run() {
                runOnUiThread { viewModelHome.setPlayersOnline() }
            }
        }, 0, 3000)

        timer.schedule(object : TimerTask() {
            override fun run() {
                runOnUiThread { viewModelHome.setWorlds() }
            }
        }, 0, 3000)

        if (RedValidator.ValidarInternet(application)) {
            threads()
        } else {
            Toast.makeText(this, "Revisa tu conexion a internet :)", Toast.LENGTH_SHORT).show()
            linearProgressIndicator.visibility = View.GONE
        }
    }

    private fun threads() {
        val executor = Executors.newFixedThreadPool(5)
        executor.execute {
            try {
                Thread.sleep(3000)
                viewModelHome.setCreatureBoss()
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
        }
        executor.execute {
            try {
                Thread.sleep(3000)
                viewModelHome.setBostedBoss()
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
        }
        executor.execute {
            try {
                Thread.sleep(3000)
                viewModelHome.setNews()
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
        }
        executor.execute {
            try {
                Thread.sleep(3000)
                viewModelHome.setNewTicker()
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
        }
        executor.execute {
            try {
                Thread.sleep(3000)
                viewModelHome.setRashirLocation()
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.nd_about) {
            val about = Intent(this, About::class.java)
            startActivity(about)
        } else if (id == R.id.nd_salir) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onClick(v: View) {}

    override fun onMenuItemClick(menuItem: MenuItem): Boolean {
        when (menuItem.itemId) {
            R.id.nd_experiencia_compartida -> {
                startActivity(Intent(this, experiencia_compartida::class.java))
                drawerLayout.close()
            }
            R.id.nd_character -> {
                startActivity(Intent(this, characters::class.java))
                drawerLayout.close()
            }
            R.id.nd_stamina -> {
                startActivity(Intent(this, Stamina::class.java))
                drawerLayout.close()
            }
            R.id.nd_mundos -> {
                startActivity(Intent(this, Mundos::class.java))
                drawerLayout.close()
            }
            R.id.nd_criaturas -> {
                startActivity(Intent(this, Criaturas::class.java))
                drawerLayout.close()
            }
            R.id.nd_guilds -> {
                startActivity(Intent(this, GuildInformation::class.java))
                drawerLayout.close()
            }
            R.id.nd_spells -> {
                startActivity(Intent(this, Spells_Tibia::class.java))
                drawerLayout.close()
            }
            R.id.nd_bless -> {
                startActivity(Intent(this, Blessings::class.java))
                drawerLayout.close()
            }
            R.id.nd_highscores -> {
                startActivity(Intent(this, Highscores::class.java))
                drawerLayout.close()
            }
            R.id.nd_houses -> {
                startActivity(Intent(this, HouseActivity::class.java))
                drawerLayout.close()
            }
            R.id.id_about -> {
                startActivity(Intent(this, About::class.java))
                drawerLayout.close()
            }
            R.id.nd_maps -> {
                startActivity(Intent(this, TibiaMaps::class.java))
                drawerLayout.close()
            }
        }
        return false
    }
}
